infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -91,33 +160,105 @@ std::vector<std::byte> read_memory(pid_t pid, unsigned long long address, size_t
+  // Rely on sanitizers/libFuzzer to produce a stacktrace by sending SIGABRT
+  // to the root process.
+  // Note: this may not be reliable or consistent if shell injection happens
+  // in an async way.
+  kill(g_root_pid, SIGABRT);
+  _exit(0);
+}
+
+void inspect_for_injection(pid_t pid, const user_regs_struct &regs) {
+  // Inspect a PID's registers for the sign of shell injection.
+  std::string path = read_string(pid, regs.rdi, kTripWire.length());
+  if (!path.length()) {
+    return;
+  }
   debug_log("inspecting");
   if (path == kTripWire) {
-    fprintf(stderr, "===BUG DETECTED: Shell injection===\n");
-    // Rely on sanitizers/libFuzzer to produce a stacktrace by sending SIGABRT
-    // to the root process.
-    // Note: this may not be reliable or consistent if shell injection happens
-    // in an async way.
-    kill(g_root_pid, SIGABRT);
-    _exit(0);
+    report_bug(kInjectionError);
   }
 }
 
+std::string get_pathname(pid_t pid, const user_regs_struct &regs) {
+  // Parse the pathname from the memory specified in the RDI register.
+  std::string raw_content = read_string(pid, regs.rdi, kShellPathnameLength);
+
+  std::string pathname = raw_content.substr(0, raw_content.find(" "));
+  debug_log("Pathname is %s (len %lu)\n", pathname.c_str(), pathname.length());
+  return pathname;
+}
+
+std::string match_shell(std::string binary_pathname) {
+  // Identify the name of the shell used in the pathname.
+  for (auto it = kShellSyntaxErrors.begin(); it != kShellSyntaxErrors.end();
+       ++it) {
+    std::string known_shell = it->first;
+    std::string binary_name = binary_pathname.substr(
+        binary_pathname.find_last_of("/") + 1, known_shell.length());
+    if (binary_name == known_shell) {
+      debug_log("Matched %s\n", binary_name.c_str());
+      return known_shell;
+    }
+  }
+  return "";
+}
+
+std::string get_shell(pid_t pid, const user_regs_struct &regs) {
+  // Get shell name used in a process.
+  std::string binary_pathname = get_pathname(pid, regs);
+  if (binary_pathname.length()) {
+    return match_shell(binary_pathname);
+  }
+  return "";
+}
+
+void match_error_pattern(std::string buffer, std::string shell) {
+  auto error_patterns = kShellSyntaxErrors.at(shell);
+  for (auto it = error_patterns.begin(); it != error_patterns.end(); ++it) {
+    debug_log("Pattern : %s\n", it->c_str());
+    debug_log("Found at: %lu\n", buffer.find(*it));
+    if (buffer.find(*it) != std::string::npos) {
+      buffer = buffer.substr(0, buffer.find("\n"));
+      printf(
+          "--- Found a sign of shell corruption ---\n"
+          "%s\n"
+          "----------------------------------------\n",
+          buffer.c_str());
+      report_bug(kCorruptionError);
+    }
+  }
+}
+
+void inspect_for_corruption(pid_t pid, const user_regs_struct &regs) {
+  // Inspect a PID's registers for shell corruption.
+  std::string buffer = read_string(pid, regs.rsi, kErrorMessageLength);
```

Comment:
Can we set the third argument to `std::min(kErrorMessageLength, regs.rdx)` so we never read more than what's expected?